src/hyp_map.rs

```diff
@@ -476,3 +504,103 @@ impl UmodeSlotMapper<'_> {
             .map_err(|_| Error::MapFailed)
     }
 }
+
+/// Umode Buffer errors.
+#[derive(Debug)]
+pub enum UmodeBufferError {
+    /// Address overflow on increment.
+    AddressOverflow,
+    /// Buffer is full.
+    NoSpace,
+}
+
+/// The U-mode buffer is a private area of the page table used by the hypervisor to pass data to the
+/// current U-mode operation. This avoids mapping hypervisor data directly in U-mode. This area is
+/// written by the hypervisor and mapped read-only in U-mode.
+pub struct UmodeBuffer {
+    /// The start address of the U-mode buffer memory.
+    start: SupervisorPageAddr,
+    /// Pointer to the unused part of the buffer.
+    addr: SupervisorPhysAddr,
```

Comment:
We wouldn't need to volatile part of it, nor the reading part of it. Feels a bit excessive for this use.

The dynamics here are a bit strange. This area is definitely fully owned by salus, and there will be really no concurrent reading. I like the use of DataInit.

This is essentially more of a rust interface question rather than architectural. I don't have particularly strong opinions about it.

Reply:
I wasn't so much worried about the concurrent reading part, but liked that it wraps the unsafe access, I sent you a patch, let me know what you think.